processor/deltatocumulativeprocessor/internal/data/add.go

```diff
@@ -23,7 +37,101 @@ func (dp Histogram) Add(in Histogram) Histogram {
 	panic("todo")
 }
 
-// nolint
 func (dp ExpHistogram) Add(in ExpHistogram) ExpHistogram {
-	panic("todo")
+	switch {
+	case dp.Timestamp() >= in.Timestamp():
```

Comment:
I think when timestamps are out of order it still makes sense to aggregate, but then to not propagate the aggregated value to the exporter. This way some disorder in timestamps would lead to a loss only in temporal resolution, but not data loss.

Reply:
correct me if I'm wrong, but I believe that would lead to incorrect data.

if we aggregated but not exported, the next sample we export would include that value, and thus be a higher increase than in reality. you would absolutely notice this when doing `rate()` on the resulting series, leading to spikes at the wrong time

Reply:
My understanding is the opposite: if we don't aggregate the late sample, we'd be forever undercounting.

Let's imagine this situation:
Events happen at timestamps 1, 2, 3 and 4. Delta values: 1, 1, 1, 1. Cumulative values: 1, 2, 3, 4.

Let's say arrival order happens to be 2, 1, 3, 4.

Collector receives the `timestamp=2, delta=1` event, aggregates it as `timestamp=2, cumulative value=1`.
Then it receives the `timestamp=1, delta=1`, aggregates it, saving `timestamp=2, cumulative value=2`.
Then it receives `timestamp=3, delta=1`, processes it into `timestamp=2, cumulative value=3`, matching the expectation.
Same with timestamp 4.

If we just throw out the `timestamp=1, delta=1` event, then the following exported samples would have value too low by 1.

Does this make sense?

Reply:
both options lose important information:

- undercounting:
  - rates: they are mostly correct. they appear at the right time and are less or equal to reality
  - absolute value: its less than reality. this value can't be relied on, unlike in cumulative systems. same happens with deltas that never make it to the collector.
- late reporting:
  - rates: they are wrong. they are offset in an unpredictable way and may peak higher than in reality.
  - absolute values: they are correct, apart from lost samples

given counters and histograms are mostly used for rate aggregations, I think it is reasonable to prefer this use-case here.
that being said, I think the out-of-order case is best solved externally, e.g. using a queue-sort processor (tbd).
if that proves insufficient, we may also consider adding `_late` metrics that record late samples. those would be useless for rates, but can bring back accuracy for total counts